### ISSUE TITLE
net.IP wrong usage

### ISSUE THREAD

**Issue by wI2L** [chihaya/chihaya#254]
Hi

I've noticed the code heavily rely on the length of `net.IP` variables to differentiate IPv4 and IPv6 protocols, and this is a mistake.
example https://github.com/chihaya/chihaya/blob/master/middleware/hooks.go#L126

Per the doc : https://golang.org/pkg/net/#IP

> Note that in this documentation, referring to an IP address as an IPv4 address or an IPv6 address is a semantic property of the address, not just the length of the byte slice: a 16-byte slice can still be an IPv4 address. 

Instead i suggest using

```go
func IsIPv4(ip net.IP) bool {
	return ip.to4() != nil
}
```
and
```go
// According to the documentation, 16-bytes addresses can
// be either protocol, but an IP returned by net.IP.To4() will
// always be an IPv4, so if the result is nil we know it is an IPv6.
func IsIPv6(ip net.IP) bool {
	return len(ip) == net.IPv6len && ip.To4() == nil
}
```

**Comment by mrd0ll4r** [comment]
Hey,

you are correct to point that out! We do rely on checking the length of an IP to determine what kind of Peer we're dealing with. We do that on purpose, mainly because it's a constant-time operation and we have to do it quite often.

In order for it to work our IPv4 addresses must have a length of 4. (And IPv6 a length of 16, but that's not really an issue.) We enforce this in the frontend (i.e. [udp/frontend.go:131-133](https://github.com/chihaya/chihaya/blob/master/frontend/udp/frontend.go#L131-L133) and [http/parser.go:82-85](https://github.com/chihaya/chihaya/blob/master/frontend/http/parser.go#L82-L85)). _We have to document this._

We should also consider not only checking for IPv4, but also testing if it is an IPv6 address otherwise (using the approach you posted). I have however never seen a `net.IP` coming from the standard library that didn't have 4 or 16 bytes.

We should also consider moving the checks from the frontend into a "sanitization" middleware that runs by default as the first hook after the frontends.

@jzelinskie input?

**Comment by wI2L** [comment]
As an approach, using a two constants to describe the protocols, and add this information to the Peer struct could be "cleaner" imo. Checks would be then very easy.

**Comment by jzelinskie** [comment]
>We should also consider moving the checks from the frontend into a "sanitization" middleware that runs by default as the first hook after the frontends.

I'd rather wait to see if people implement new frontends before we start catering to an invisible market. No point adding meaningless overhead to all requests.

>As an approach, using a two constants to describe the protocols, and add this information to the Peer struct could be "cleaner" imo. Checks would be then very easy.

Can you expand on this a bit further? When the tracker is in use, peers make up most of the memory, so I'd be hesitant to add more fields to the Peer struct unless it was necessary.

**Comment by wI2L** [comment]
@jzelinskie A uint8 would be enough to store 0 or 1 to distinguish between IPv4 and  IPv6. This would require 1 extra byte when serialized to `serializedPeer`.

**Comment by mrd0ll4r** [comment]
We don't need to save that information in a `serializedPeer`, we can just recompute it on retrieval.

**Comment by wI2L** [comment]
Or that indeed.

2016-11-17 18:17 GMT+01:00 mrd0ll4r <notifications@github.com>:

**Comment by mrd0ll4r** [comment]
> I'd rather wait to see if people implement new frontends before we start catering to an invisible market. No point adding meaningless overhead to all requests.

We don't actually know how high the overhead of one more hook is. It's a cleaner solution than hiding this in the frontends, regardless of whether there will be more frontends.

**Comment by mrd0ll4r** [comment]
I created PR #258 to illustrate my ideas and show some benchmarks, comments welcome :)

I believe it's much cleaner to do checks like the IP sanitation or the `numWant` bounds check in a middleware. I'd also be in favor of adding something like `defaultNumWant` to the Hook, which sets numWant to that value in case it was specified as zero.

(For an example of why I want the numWant check: the `optmem` storage implementation I wrote did not check numWant by default and thus failed.)

**Comment by mrd0ll4r** [comment]
@jzelinskie should we move forward with the sanitation hook and checks?

**Comment by jzelinskie** [comment]
benchmarks prove it seems like a pretty fair tradeoff.

**Comment by mrd0ll4r** [comment]
This has been fixed in #258
